### PR TITLE
BugFix: Decomposed estimation time

### LINKED ISSUE #198: Decomposed estimation time
**NAME:** Bug report
**ABOUT:** Decomposed estimation time
**ASSIGNEES:** Denis Jorge Gandarillas Delgado

**Describe the bug**
When you view the routes, the time estimate is decomposed, because it first shows you the actual estimate, but then changes to a negative estimate.

**To Reproduce**
1. Login to the application.
2. Register or log in if you have not already done so.
3. Select a point on the map.
4. Select whether you want to run the route from your current position or from another selected position on the map.
5. In case you choose another position, select the starting position on the map.
6. Select "Start".
7. Display the estimated time for each route.

**Expected behavior**
It should show a correct time estimate.

**Screenshots**
image

**Smartphone:**
 - Device: Realme C11 2021
 - OS: Android
 - Version: 11

### PR DESCRIPTION
**Pull request by zrock-dev**
fixes #198

### CODE REVIEW

**Review by aguirrekarina — APPROVED** [review]

**Review by Elvis-Castro-H — COMMENTED** [review]
Approved

**Review by Elvis-Castro-H — APPROVED** [review]
Approved

**Comment by zrock-dev** [comment]
Marked to draft due to instability with the estimation of the time and distance.

File: app/src/main/java/com/isc/hermes/controller/InfoRouteController.java [hunk 1]
```diff
@@ -140,6 +140,16 @@ private void setActionButtons() {
         });
     }
 
+    private void stopNavigationMode(){
```
Comment (daniel-espinoza-79) [inline app/src/main/java/com/isc/hermes/controller/InfoRouteController.java:143]: Javadoc missing

File: app/src/main/java/com/isc/hermes/controller/NavigationOptionsController.java [hunk 2]
```diff
@@ -474,15 +472,12 @@ public MapWayPointController getMapWayPointController() {
      * drawPolylineEverySecond method.
      */
     private void updateNavigatedRoute() {
+        polylineRouteUpdaterController = polylineRouteUpdaterController == null ? new PolylineRouteUpdaterController(startPoint,finalPoint) :
+        polylineRouteUpdaterController;
         polylineRouteUpdaterController.drawPolylineEverySecond();
     }
 
     private void startRouteEstimationManager(String JSONRoute){
```
Comment (daniel-espinoza-79) [inline app/src/main/java/com/isc/hermes/controller/NavigationOptionsController.java:480]: Javadoc missing

File: app/src/main/java/com/isc/hermes/model/navigation/UserRouteTracker.java [hunk 3]
```diff
@@ -137,4 +154,34 @@ public boolean isUserInSegment(LatLng userLocation) {
...
+            trackRecoveryProtocol.attemptDeepRecovery();
+        } else {
+            trackRecoveryProtocol.attemptSimpleRecovery(routeSegmentIndex);
+        }
+
+        if (trackRecoveryProtocol.isAttemptSuccessful()) {
+            currentSegment = trackRecoveryProtocol.getTrack();
+            routeSegmentIndex = routeSegments.indexOf(currentSegment);
+            routeDistanceHandler.update(routeSegmentIndex);
+            System.out.printf("User on track #%d\n", routeSegmentIndex);
```
Comment (daniel-espinoza-79) [inline app/src/main/java/com/isc/hermes/model/navigation/UserRouteTracker.java]: should implement log statements (Log.i for information and Log.e for error) to provide more meaningful messages. The use of TAG allows easy filtering and identification of log messages in Logcat.

File: app/src/main/java/com/isc/hermes/model/navigation/TrackRecoveryProtocol.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    public void attemptDeepRecovery() {
+        for (int index = 0; index < routeSegments.size(); index++) {
+            System.out.printf("ATTEMPT RECOVERY ON TRACK #%d\n", index);
+            attemptSimpleRecovery(index);
+            if (isAttemptSuccessful()) {
+                break;
+            } else {
+                System.out.println("ATTEMPT FAILED");
```
Comment (daniel-espinoza-79) [inline app/src/main/java/com/isc/hermes/model/navigation/TrackRecoveryProtocol.java]: should implement log statements (Log.i for information and Log.e for error) to provide more meaningful messages.

File: app/src/main/java/com/isc/hermes/model/navigation/TrackRecoveryProtocol.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        currentLocation = CurrentLocationModel.getInstance();
+        foundedTrack = null;
+    }
+
+    public boolean isAttemptSuccessful() {
+        return (foundedTrack != null);
+    }
+
+    public void attemptSimpleRecovery(int routeSegmentIndex) {
+        if (routeSegments.size() == 1) {
```
Comment (daniel-espinoza-79) [inline app/src/main/java/com/isc/hermes/model/navigation/TrackRecoveryProtocol.java]: it should be a static data, there is no context that it is 1

File: app/src/main/java/com/isc/hermes/model/navigation/TrackRecoveryProtocol.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            simpleStrategy();
+        } else if (routeSegments.size() == 2) {
```
Comment (daniel-espinoza-79) [inline app/src/main/java/com/isc/hermes/model/navigation/TrackRecoveryProtocol.java]: it should be a static data, there is no context that it is 2

File: app/src/main/java/com/isc/hermes/model/navigation/LiveRouteEstimationsWorker.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        }
+
+        updateEstimatedArrivalTime(totalEstimatedDistance);
+    }
+
+    /**
+     * Updates the estimated arrival time based on the total estimated distance and transportation type.
+     */
+    private void updateEstimatedArrivalTime(double distance) {
+        int totalEstimatedArrivalTime = (int) Math.ceil(((distance / transportationType.getVelocity()) * 60));
```
Comment (daniel-espinoza-79) [inline app/src/main/java/com/isc/hermes/model/navigation/LiveRouteEstimationsWorker.java:60]: it should be a static data, there is no context that it is 60

**Review by daniel-espinoza-79 — APPROVED** [review]

**Pull request merged**